### PR TITLE
router: retry overloaded requests

### LINKED ISSUE #10706: Perform retries even when x-envoy-overloaded is set
The current handling of x-envoy-overloaded disables retries whenever this header is returned from the upstream. As a result, requests that fail due to circuit breakers on the ingress side in a mesh setup can result in the downstream request failing without any retries being attempted simply because one of the upstream hosts was overloaded. 

We should consider removing this behavior, instead recommending configuring retry circuit breakers on egress to prevent retry storms.

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: Retries overloaded requests
Risk Level: Low
Testing: Added
Docs Changes: Updated
Release Notes: Added

Fixes #10706

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@snowp @mattklein123 PTAL

**Comment by mattklein123** [comment]
@tonya11en can you take a look at the linked issue and do a first pass over this? I want to make sure we are all in agreement here on the best overall approach.

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 1]
```diff
@@ -122,8 +122,9 @@ headers <config_http_filters_router_headers_consumed>`. The following configurat
   :ref:`retry priority <envoy_api_field_route.RetryPolicy.retry_priority>` can be
   configured to adjust the priority load used when selecting a priority for retries.
 
-Note that retries may be disabled depending on the contents of the :ref:`x-envoy-overloaded
-<config_http_filters_router_x-envoy-overloaded_consumed>`.
+Note that Envoy retries requests when :ref:`x-envoy-overloaded
+<config_http_filters_router_x-envoy-overloaded_set>` is present. It is recommended to configure
```
Comment (tonya11en) [inline docs/root/intro/arch_overview/http/http_routing.rst:126]: This is already set to a default value of 3, so folks should already be protected. I also would mention retry budgets here instead of the max_retries CB threshold. Especially since the max_retries circuit breaker already is enabled by default to a low value. This change will certainly result in a dramatic increase in retry overflows when circuit breaking occurs.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -12,6 +12,7 @@ Changes
   Disabled by default and can be enabled via :ref:`enable_upstream_stats <envoy_v3_api_field_extensions.filters.http.grpc_stats.v3.FilterConfig.enable_upstream_stats>`.
 * http: fixed a bug where the upgrade header was not cleared on responses to non-upgrade requests.
   Can be reverted temporarily by setting runtime feature `envoy.reloadable_features.fix_upgrade_response` to false.
+* router: added support for retrying requests when upstream responds with :ref:`x-envoy-overloaded <config_http_filters_router_x-envoy-overloaded_set>`.
```
Comment (tonya11en) [inline docs/root/version_history/current.rst:15]: I would reword this to be more clear about what this change actually does. Maybe say something like:

> router: allow retries by default when encountering x-envoy-overloaded header

As it reads now, it can imply that the old behavior is still intact.

File: docs/root/intro/arch_overview/http/http_routing.rst (hunk 1 above)
Comment (mattklein123) [reply docs/root/intro/arch_overview/http/http_routing.rst:126]: As an aside, I really wonder if we should at some point turn on the budgets by default vs. the hard limit. It's a much better default?

File: docs/root/configuration/http/http_filters/router_filter.rst [hunk 3]
```diff
@@ -274,14 +274,6 @@ for the next health check interval. The host can become healthy again via standa
 checks. See the :ref:`health checking overview <arch_overview_health_checking>` for more
 information.
 
-.. _config_http_filters_router_x-envoy-overloaded_consumed:
-
-x-envoy-overloaded
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/router_filter.rst:279]: Is this header documented anywhere else? I think we still should document it. I think it does still make sense to send it as downstream may want to track it, but we should document somewhere?

Comment (tonya11en) [reply docs/root/configuration/http/http_filters/router_filter.rst:279]: I checked earlier- it's explained further down in the same file.

File: docs/root/version_history/current.rst (hunk 2 above)
Comment (snowp) [reply docs/root/version_history/current.rst:15]: might be good to explicitly mention circuit breakers, readers might not be aware of this header

Comment (ramaraochavali) [reply docs/root/version_history/current.rst:15]: I updated the doc and mentioned about using circuit breaker as a recommendation in the retry section. Do you think we should add it in the Version History as well?

File: docs/root/intro/arch_overview/http/http_routing.rst (hunk 1 above)
Comment (ramaraochavali) [reply docs/root/intro/arch_overview/http/http_routing.rst:126]: @tonya11en  I will adjust the doc.

@mattklein123 Agreed. Retry budgets are better defaults than hard limits. We should consider doing that

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 4]
```diff
@@ -122,8 +122,10 @@ headers <config_http_filters_router_headers_consumed>`. The following configurat
   :ref:`retry priority <envoy_api_field_route.RetryPolicy.retry_priority>` can be
   configured to adjust the priority load used when selecting a priority for retries.
 
-Note that retries may be disabled depending on the contents of the :ref:`x-envoy-overloaded
-<config_http_filters_router_x-envoy-overloaded_consumed>`.
+Note that Envoy retries requests when :ref:`x-envoy-overloaded
+<config_http_filters_router_x-envoy-overloaded_set>` is present. It is recommended to either configure
+:ref:`retry budgets (preferred) <envoy_api_field_cluster.CircuitBreakers.Thresholds.retry_budget>` or set
+:ref:`maximum active retries circuit breaker <arch_overview_circuit_break>` to an appropriate value avoid retry storms.
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/http/http_routing.rst:128]:
```suggestion
:ref:`maximum active retries circuit breaker <arch_overview_circuit_break>` to an appropriate value to avoid retry storms.
```

File: source/common/router/retry_state_impl.cc [hunk 5]
```diff
@@ -281,10 +281,6 @@ RetryStatus RetryStateImpl::shouldHedgeRetryPerTryTimeout(DoRetryCallback callba
 }
 
 bool RetryStateImpl::wouldRetryFromHeaders(const Http::ResponseHeaderMap& response_headers) {
-  if (response_headers.EnvoyOverloaded() != nullptr) {
```
Comment (mattklein123) [inline source/common/router/retry_state_impl.cc:284]: Can you remove this from the O(1) header map? I think we don't need this anymore as the header can be added by string reference where it is added currently?
